sol/main.c

```diff
@@ -0,0 +1,513 @@
+		&req_queue->tb_period);
+
+	rte_eth_link_get(sol_conf->net->back.id, &link);
+	req_queue->rate = (uint64_t)link.link_speed * 1000 * 1000 / 8;
+	cycles_per_byte = (rte_get_tsc_hz() << RTE_SCHED_TIME_SHIFT)
+		/ req_queue->rate;
+	req_queue->inv_cycles_per_byte = rte_reciprocal_value(cycles_per_byte);
+
+	req_queue->tb_size = req_queue->rate * REQ_QUEUE_BW_RATE;
+	/* The DPDK sched library starts the bucket half full. */
+	req_queue->tb_credits = DEFAULT_TB_SIZE / 2;
+	req_queue->tb_time = 0;
+
+	/* Timing. */
+	req_queue->time_cpu_cycles = rte_get_tsc_cycles();
+	req_queue->time_cpu_bytes = 0;
+	req_queue->time = 0;
+
+	req_queue->head = NULL;
+	req_queue->bmp_array = req_queue->memory;
+
+	bmp_mem_size =
+		rte_bitmap_get_memory_footprint(req_queue->num_priorities);
+	req_queue->bmp = rte_bitmap_init(req_queue->num_priorities,
+		req_queue->bmp_array, bmp_mem_size);
+	if (req_queue->bmp == NULL) {
+		RTE_LOG(ERR, GATEKEEPER, "sol: can't initialize req bitmap\n");
+		rte_free(req_queue);
+		return NULL;
+	}
+
+	return req_queue;
+}
+
+static int
+cleanup_sol(struct sol_config *sol_conf)
+{
+	destroy_mailbox(&sol_conf->mb);
+	rte_free(sol_conf->req_queue);
+	sol_conf->net = NULL;
+	gk_conf_put(sol_conf->gk);
+	sol_conf->gk = NULL;
+	rte_free(sol_conf);
+	return 0;
+}
+
+static int
+sol_proc(void *arg)
+{
+	struct sol_config *sol_conf = (struct sol_config *)arg;
+	unsigned int lcore = sol_conf->lcore_id;
+	uint8_t tx_port_back = sol_conf->net->back.id;
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"sol: the Solicitor block is running at lcore = %u\n", lcore);
+
+	while (likely(!exiting)) {
+		enqueue_reqs(sol_conf);
+		dequeue_reqs(sol_conf, tx_port_back,
+			sol_conf->tx_queue_back);
+	}
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"sol: the Solicitor block at lcore = %u is exiting\n", lcore);
+
+	return cleanup_sol(sol_conf);
+}
+
+static int
+sol_state1(void *arg)
+{
+	struct sol_config *sol_conf = arg;
+	int ret;
+
+	ret = get_queue_id(&sol_conf->net->back, QUEUE_TYPE_TX,
+		sol_conf->lcore_id);
+	if (ret < 0) {
+		RTE_LOG(ERR, GATEKEEPER, "sol: cannot assign a TX queue for the back interface for lcore %u\n",
+			sol_conf->lcore_id);
+		goto cleanup;
+	}
+	sol_conf->tx_queue_back = ret;
+
+	sol_conf->req_queue = req_queue_init();
+	if (sol_conf->req_queue == NULL) {
+		ret = -1;
+		goto cleanup;
+	}
+
+	ret = init_mailbox("gk_reqs", MAILBOX_MAX_ENTRIES,
+		sizeof(struct priority_req), sol_conf->lcore_id,
+		&sol_conf->mb);
+	if (ret < 0)
+		goto cleanup;
+
+	return 0;
+
+cleanup:
+	cleanup_sol(sol_conf);
+	return -1;
+}
+
+int
+run_sol(struct net_config *net_conf, struct gk_config *gk_conf,
+	struct sol_config *sol_conf)
+{
+	int ret;
+
+	if (net_conf == NULL || gk_conf == NULL || sol_conf == NULL) {
+		ret = -1;
+		goto out;
+	}
+
+	if (!net_conf->back_iface_enabled) {
+		RTE_LOG(ERR, GATEKEEPER, "sol: back interface is required\n");
+		ret = -1;
+		goto out;
+	}
+
+	ret = net_launch_at_stage1(net_conf, 0, 0, 1, 0, sol_state1, sol_conf);
```

Comment:
`sol_stage1()` won't run until sometime after `run_sol()` is finished, so I think it should be OK to set `sol_conf->net` later in this function.